packages/medusa/src/services/order.js

```diff
@@ -0,0 +1,447 @@
+    const { value, error } = Validator.address().validate(address)
+    if (error) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        "The address is not valid"
+      )
+    }
+
+    return value
+  }
+
+  /**
+   * Used to validate email.
+   * @param {string} email - the email to vaildate
+   * @return {string} the validate email
+   */
+  validateEmail_(email) {
+    const schema = Validator.string().email()
+    const { value, error } = schema.validate(email)
+    if (error) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_ARGUMENT,
+        "The email is not valid"
+      )
+    }
+
+    return value
+  }
+
+  async partitionItems_(shipping_methods, items) {
+    let updatedMethods = []
+    // partition order items to their dedicated shipping method
+    await Promise.all(
+      shipping_methods.map(async method => {
+        const { profile_id } = method
+        const profile = await this.shippingProfileService_.retrieve(profile_id)
+        // for each method find the items in the order, that are associated
+        // with the profile on the current shipping method
+        method.items = items.filter(({ content }) => {
+          if (Array.isArray(content)) {
+            // we require bundles to have same shipping method, therefore:
+            return profile.products.includes(content[0].product._id)
+          } else {
+            return profile.products.includes(content.product._id)
+          }
+        })
+        updatedMethods.push(method)
+      })
+    )
+    return updatedMethods
+  }
+
+  /**
+   * Gets an order by id.
+   * @param {string} orderId - id of order to retrieve
+   * @return {Promise<Order>} the order document
+   */
+  async retrieve(orderId) {
+    const validatedId = this.validateId_(orderId)
+    const order = await this.orderModel_
+      .findOne({ _id: validatedId })
+      .catch(err => {
+        throw new MedusaError(MedusaError.Types.DB_ERROR, err.message)
+      })
+
+    if (!order) {
+      throw new MedusaError(
+        MedusaError.Types.NOT_FOUND,
+        `Order with ${orderId} was not found`
+      )
+    }
+    return order
+  }
+
+  /**
+   * Creates an order
+   * @param {object} order - the order to create
+   * @return {Promise} resolves to the creation result.
+   */
+  async create(order) {
+    return this.orderModel_.create(order).catch(err => {
+      throw new MedusaError(MedusaError.Types.DB_ERROR, err.message)
+    })
+  }
+
+  /**
+   * Updates an order. Metadata updates should
+   * use dedicated method, e.g. `setMetadata` etc. The function
+   * will throw errors if metadata updates are attempted.
+   * @param {string} orderId - the id of the order. Must be a string that
+   *   can be casted to an ObjectId
+   * @param {object} update - an object with the update values.
+   * @return {Promise} resolves to the update result.
+   */
+  async update(orderId, update) {
+    const validatedId = this.validateId_(orderId)
```

Comment:
We already do this in `retrieve` so there is no need to do it here as well

Reply:
But we don't use `retrieve` when updating, so you could in theory pass `junk-id`. Should we add a `retrieve` call, such that we check the existing of the order before updating it?

Reply:
There is a retrieve call 2 lines below this one, which is used to check if the current status of the order allows updates :D

Reply:
There is a retrieve call a couple of lines below - it's used to check if the current status of the order allows updates.

Reply:
Waow... I wonder where I was looking. My bad big time haha 😄